text/0001-workspace-consistent-versions.md

```diff
@@ -0,0 +1,572 @@
+        }
+      },
+      "enforceConsistencyTransitively": true,
+      "allowTransitiveMismatches": ["object-assign"]
+    }
+  }
+}
+```
+
+`enforceConsistencyTransitively` will error if there is ever more than one item in the `packages` block of `pnpm-lock.yaml` for a dependency, even duplicates on the same version but appearing multiple times due to peer dependencies.
+
+The `allowTransitiveMismatches` option will provide an escape hatch if `enforceConsistencyTransitively` is enabled and a dependency requires a dependency version outside of the `workspaceConistent` range. For example, suppose all workspace packages specify `workspace-consistent:*` on the `object-assign` ponyfill, but `react` depends on `object-assign@^4.1.1`. In this scenario, two copies of `object-assign` across different major versions (3.x and 4.x) are present in the dependency graph. Adding `object-assign` to the `allowTransitiveMismatches` array marks this as acceptable. While the multiple copies of `object-assign` are non-ideal, this escape hatch allows users to selectively relax the `enforceConsistencyTransitively` check without disabling it entirely.
+
+### Allowed Deviations
+
+The default behavior is to show an error if a workspace package does not use `workspace-consistent:` for a configured dependency. However, sometimes a workspace package may not be able to use the same version of a dependency as the other packages in the monorepo. The `allowedDeviations` option can be used as an escape hatch.
+
+Once specified, `@monorepo/baz` will be allowed to declare a version range that does not begin with `workspace-consistent:` for `@types/react`.
+
+```json5
+{
+  "name": "some-monorepo",
+  "pnpm": {
+    "workspaceConsistent": {
+      "groups": {
+        "default": {
+          "@types/react": "^17.0.43",
+          "react": "^17.0.2",
+          "react-dom": "^17.0.2",
+        }
+      },
+      "allowedDeviations": {
+        "@types/react": ["@monorepo/baz"]
+      },
+    }
+  }
+}
+```
+
+### Version Groups
+
+Users may want to declare different subset partitions of consistent version sharing. Borrowing the phrase [_version groups_ from syncpack](https://github.com/JamieMason/syncpack#versiongroups), this could be defined as such:
+
+```json5
+{
+  "name": "example-monorepo",
+  "private": true,
+  "pnpm": {
+    "workspaceConsistent": {
+      "groups": {
+        "default": {
+          "react": "^17.0.2",
+          "react-dom": "^17.0.2",
+          "redux": "^4.1.2"
+        },
+        "react-rc": {
+          "react": "^18.0.0-rc.3",
+          "react-dom": "18.0.0-rc.3"
+        },
+        "react-next": {
+          "react": "18.0.0-rc.3-next-033fe52b4-20220325",
+          "react-dom": "18.0.0-rc.3-next-033fe52b4-20220325"
+        },
+      }
+    }
+  }
+}
+```
+
+Workspace packages would then use `workspace-consistent:<version-group>` as the protocol. Simply `workspace-consistent:` would refer to the `default` group.
```

Comment:
This was an intentional deviation from how `syncpack` manages version groups, but I'm rethinking this decision. Namely, the proposal here makes packages to refer to a version group. `syncpack` goes the other way and requires a version group's config to specify a list of what packages are included.

With the current proposal, I like that:

- You can look at `workspace-consistent:<version-group>` and get a rough idea of what group the package is a part of.
- Allowing packages to refer to version groups makes them many-to-many. You can define that a package is part of the Jest RC version group and the React RC version group. `syncpack` requires version groups to be partitions of the workspace. A package can't be a part of more than one version group in `syncpack`.

However:

- Making version groups partitions (ala `syncpack`) is valuable for the `enforceConsistencyTransitively` option. Suppose `react-rc` and `react-next` version groups are defined. As written the semantics of making sure a `react` RC version doesn't leak into the transitive dependency graph for workspace packages in the `react-next` group isn't clear.

I don't have significant planned change to the rest of this document, but going to think about this section a bit more.

Reply:
I don't know that there's much you can do to avoid version leakage beyond having the order in which your version groups are defined matter and have the earliest take priority when multiple can satisfy a given package's dependency specifier. When it matters, the packages in question should be using peerDependencies so that the version can be controlled by the host.